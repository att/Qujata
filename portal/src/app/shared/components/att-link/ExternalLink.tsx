import { BaseLinkProps, LinkRel, LinkTarget } from './Link.model';
import { getLinkClassName } from './Link.util';

export interface ExternalLinkProps extends BaseLinkProps {
  target?: LinkTarget;
  rel?: LinkRel;
}

export const ExternalLink: React.FC<ExternalLinkProps> = (props: ExternalLinkProps) => (
  <a
    href={props.link}
    target={props.target || LinkTarget.SELF}
    rel={props.rel || LinkRel.NONE}
    className={getLinkClassName(props)}
    onClick={props.onLinkClick}
    title={props.linkTitle}
    data-cy={props.e2eId}
  >
    {props.children}
  </a>
);
