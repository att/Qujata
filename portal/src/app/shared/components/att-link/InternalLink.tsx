import { useMemo } from 'react';
import { Link as RouterLink, Path, useResolvedPath } from 'react-router-dom';
import { BaseLinkProps } from './Link.model';
import { getLinkClassName } from './Link.util';

export interface InternalLinkProps extends BaseLinkProps {
  queryParams?: string;
  replaceUrl?: boolean;
}

export const InternalLink: React.FC<InternalLinkProps> = (props: InternalLinkProps) => {
  const pathname: string = useFixReactRouterRelativePathname(props.link);
  return (
    <RouterLink
      to={{
        pathname,
        search: props.queryParams,
      }}
      replace={props.replaceUrl}
      className={getLinkClassName(props)}
      onClick={props.onLinkClick}
      title={props.linkTitle}
      data-cy={props.e2eId}
    >
      {props.children}
    </RouterLink>
  );
};

/*
  https://reactrouter.com/docs/en/v6/upgrading/v5#note-on-link-to-values
  react router Link component support link back "up" to parent routes, using a leading '..' segment
  From some reason it's working fine when "to" is a string but if it's an object it's not working
  To solve it, we will resolve the path here and send the absolute  path to Link component
*/
function useFixReactRouterRelativePathname(pathname: string): string {
  const memoizedTo: Partial<Path> = useMemo<Partial<Path>>(() => ({ pathname }), [pathname]);
  return useResolvedPath(memoizedTo).pathname;
}
