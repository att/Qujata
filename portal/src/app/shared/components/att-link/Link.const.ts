import styles from './Link.module.scss';
import { LinkSize, LinkStyle } from './Link.model';

export const LinkCssClassByStyleType: Map<LinkStyle, string> = new Map<LinkStyle, string>([
  [LinkStyle.PRIMARY, styles.primary_link],
  [LinkStyle.SECONDARY, styles.secondary_link],
  [LinkStyle.WRAPPER, styles.wrapper_link],
  [LinkStyle.TEXT, styles.text_link],
  [LinkStyle.NONE, ''],
]);

export const LinkCssClassBySize: Map<LinkSize, string> = new Map<LinkSize, string>([
  [LinkSize.SMALL, styles.sm_link],
  [LinkSize.MEDIUM, styles.md_link],
  [LinkSize.LARGE, styles.lg_link],
  [LinkSize.NONE, ''],
]);
