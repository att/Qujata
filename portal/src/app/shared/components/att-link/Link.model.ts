export interface BaseLinkProps {
  link: string;
  linkTitle?: string;
  styleType: LinkStyle;
  size: LinkSize;
  className?: string;
  children: React.ReactNode;
  onLinkClick?: (e: React.MouseEvent<HTMLAnchorElement>) => void;
  e2eId?: string;
}

interface IBaseLink {
  link: string;
  linkTitle?: string;
  styleType: LinkStyle;
  size: LinkSize;
  className?: string;
  e2eId?: string;
  target?: LinkTarget;
  rel?: LinkRel;
  text: string;
}

export interface IExternalLink extends IBaseLink {
  target?: LinkTarget;
  rel?: LinkRel;
}

export enum LinkStyle {
  PRIMARY,
  SECONDARY,
  WRAPPER,
  TEXT,
  NONE,
}

export enum LinkSize {
  SMALL,
  MEDIUM,
  LARGE,
  NONE,
}

export enum LinkTarget {
  BLANK = '_blank',
  SELF = '_self',
  PARENT = '_parent',
  TOP = '_top',
  FRAMENAME = 'framename',
}

export enum LinkRel {
  ALTERNATE = 'alternate',
  AUTHOR = 'author',
  BOOKMARK = 'bookmark',
  EXTERNAL = 'external',
  HELP = 'help',
  LICENSE = 'license',
  NEXT = 'next',
  NO_OPENER = 'noopener',
  NOFOLLOW = 'nofollow',
  NONE = '',
  NOREFERRER = 'noreferrer',
  PREV = 'prev',
  SEARCH = 'search',
  TAG = 'tag',
}
