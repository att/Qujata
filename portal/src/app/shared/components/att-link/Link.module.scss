@import './../att-button/Button.reference';

%linkDefinitions {
  align-items: center;
  display: flex;
  justify-content: center;
  text-decoration: none;
  
  @extend %basicButtonStyle;
}

.sm_link {
  @extend %smallSize;
}

.md_link {
  @extend %mediumSize;
}

.lg_link {
  @extend %largeSize;
}

.primary_link {
  @extend %linkDefinitions;
  @extend %primaryStyle;
}

.secondary_link {
  @extend %linkDefinitions;
  @extend %secondaryStyle;
}

.wrapper_link {
  text-decoration: none;
  color: var($primaryWhite);
}

.text_link {
  &:not(:hover) {
    color: var($primaryWhite);
    text-decoration: none;
  }

  &:hover {
    color: var($primaryWhite);
    text-decoration: underline;
  }
}