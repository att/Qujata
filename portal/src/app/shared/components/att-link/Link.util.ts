import cn from 'classnames';
import { LinkCssClassBySize, LinkCssClassByStyleType } from './Link.const';
import { BaseLinkProps } from './Link.model';

export function getLinkClassName({ className, size, styleType }: BaseLinkProps) {
  return cn(className, LinkCssClassBySize.get(size), LinkCssClassByStyleType.get(styleType));
}
