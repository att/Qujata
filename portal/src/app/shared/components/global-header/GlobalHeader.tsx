import { PropsWithChildren } from 'react';
import cn from 'classnames';
import { NAVIGATION_ROUTES } from '../../../routes-navigation.const';
import { InternalLink, LinkStyle, LinkSize } from '../att-link';
import styles from './GlobalHeader.module.scss';
import QujataLogoSvg from '../../../../assets/images/qujata-logo.svg';
import { INavigationTab, NavigationTab } from '../navigation-tab/NavigationTab';

export interface GlobalHeaderProps {
  tabs: INavigationTab[];
  className?: string
}
export const GlobalHeader: React.FC<PropsWithChildren<GlobalHeaderProps>> = (props: PropsWithChildren<GlobalHeaderProps>) => (
  <header className={cn(props.className, styles.global_header)}>
    <div className={styles.header_left_block}>
      <InternalLink
        className={styles.logo}
        link={NAVIGATION_ROUTES.home}
        styleType={LinkStyle.WRAPPER}
        size={LinkSize.NONE}
      >
        <img className={styles.home_icon} src={QujataLogoSvg} alt="globe" />
      </InternalLink>
    </div>
    <NavigationTab tabs={props.tabs} />
  </header>
);
