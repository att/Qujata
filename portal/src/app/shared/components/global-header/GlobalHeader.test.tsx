import { render, RenderResult } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { InternalLink } from '../att-link';
import { GlobalHeader } from './GlobalHeader';

jest.mock('../att-link');
describe('GlobalHeader', () => {
  test('renders global header ', () => {
    (InternalLink as jest.Mock).mockImplementation(() => <div>InternalLink</div>);
    const { container }: RenderResult = render(<GlobalHeader title='react' />, { wrapper: MemoryRouter });
    expect(container.firstChild).toMatchSnapshot();
  });
});
