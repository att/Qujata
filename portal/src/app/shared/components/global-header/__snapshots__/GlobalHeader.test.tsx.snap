// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GlobalHeader renders global header  1`] = `
<header
  class="global_header"
>
  <div
    class="header_left_block"
  >
    <div>
      InternalLink
    </div>
  </div>
  <div
    class="tabs"
  >
    <div>
      <a
        aria-current="page"
        class="tab activeTab active"
        href="/"
      >
        Home
      </a>
    </div>
    <div>
      <a
        class="tab"
        href="/all-experiments"
      >
        All Experiments
      </a>
    </div>
  </div>
</header>
`;
